design/ceph-osd-removal.md

```diff
@@ -0,0 +1,113 @@
+# Removing OSDs from a Ceph cluster
+
+**Targeted for v1.X**
+
+## Background
+
+Removing OSDs is a common day-2 operation and users need to be able to perform
+this regularly.
+
+With the effort to add OSD replacement in the Ceph Dashboard, there is a demand
+for automating the deletion of individual OSDs.
+
+## Status
+
+As of Rook 1.1, there is some preliminary support for removing OSDs:
+
+### Removing Nodes
+
+Rook fully supports the removal of whole nodes from a Cluster, including all
+OSDs. This implementation does every call needed to completely remove all OSDs
+on that specific node from a Ceph cluster if that cluster no longer appears in
+the list of nodes in the CephCluster CR.
+
+There are concerns that this implementation might be dangerous, if a node is
+being removed inadvertently.
+
+### Removing individual OSDs
+
+Rook partially supports individual OSD-removal: The Rook operator monitors Ceph
+health and initiates the remove of OSDs that are, according to
+`ceph safe-to-destroy`, safe to be destroyed. Rook then removes the OSD
+Deployment automatically.
+
+Note that the device might stay referenced within the CephCluster CR.
+
+This feature is mainly aimed at automatically removing unresponsive OSDs.
+
+### Manual removal
+
+There is an existing Rook issue
+[#1827](https://github.com/rook/rook/issues/1827) containing a description of
+how to remove OSDs manually.
+
+## Design considerations
+
+Any future implementation within Rook needs to fulfill some requirements:
+
+1. To prevent inadvertent removal of OSDs, the act of removing OSDs needs to be
+   executed explicitly and manually.
+2. Omitting node or drive declarations within the CephCluster CR should not
+   physically remove OSDs.
+3. Except for physically removing data on disks, there should not be any manual
+   step required to allow a subsequent creation of a replacement OSD on the same
+   node.
+4. There should not be a perceived fundamental difference when removing whole
+   nodes versus removing single OSDs form a node.
+
+## Algorithm Outline
+
+Removing OSDs looks as follows when looking at the bare minimum of required
+steps (pseudocode):
+
+```python
```

Comment:
I don't think it's a good idea to include rough code snippet into a design doc, the design must be understandable by people not familiar with programming. That's the reasons why we write text into a design and then implement separately the code.

It's probably better to break down the code snippet into a bullet list.

Reply:
Hm. We need to agree on a very specific algorithm here and not a rough idea. To express my ideas, I personally would rather favorite with some clear pseudo-code compared to some obfuscated prose. 

We have two audiences here:

1. us deciding on one specific distributed algorithm
2. Some reader who wants to understand some backgrounds in the future.

As we're talking about some Rook intricacies here, I'd vote for making (1) a priority. And take (2) as an afterthought.

Reply:
Still, any programmer can translate a bullet list into an algorithm. Rook intricacies? I think it's slightly overstating, I only see a loop and a few conditions, no rocket science algorithm... Anyway, if I'm the only one thinking this way I won't hold the PR.

Reply:
> I'm the only one thinking this way I won't hold the PR.

Would converting this into bullet list work for you, after we've agreed on something?

Reply:
Yes definitely if the others don't see the need then let's keep it this way otherwise convert to bullet list :)